signing/signing.go

```diff
@@ -0,0 +1,324 @@
+	"github.com/sigstore/cosign/cmd/cosign/cli/rekor"
+	"github.com/sigstore/cosign/pkg/cosign"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/types"
+	hashedrekord "github.com/sigstore/rekor/pkg/types/hashedrekord/v0.0.1"
+	rekord "github.com/sigstore/rekor/pkg/types/rekord/v0.0.1"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+)
+
+type ScorecardOutput struct {
+	SarifOutput string
+	JsonOutput  string
+}
+
+// verifySignature receives the scorecard analysis payload, looks up its associated tlog entry and
+// certificate, and extracts the repository's workflow file to ensure its legitimacy.
+func VerifySignature(w http.ResponseWriter, r *http.Request) {
+	ctx := context.Background()
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "error reading http request body", http.StatusInternalServerError)
+		log.Println(err)
+		return
+	}
+
+	// Unmarshal request body.
+	var scorecardOutput ScorecardOutput
+	err = json.Unmarshal(reqBody, &scorecardOutput)
+	if err != nil {
+		http.Error(w, "error unmarshalling request body", http.StatusInternalServerError)
+		log.Println(err)
+		return
+	}
+
+	// Lookup results payload to get the repo info from the corresponding entry & cert.
+	sarifRepoPath, sarifRepoRef, err1 := lookupPayload(ctx, []byte(scorecardOutput.SarifOutput))
+	jsonRepoPath, jsonRepoRef, err2 := lookupPayload(ctx, []byte(scorecardOutput.JsonOutput))
+	if err1 != nil || err2 != nil {
+		http.Error(w, "error looking up payloads and processing/verifying entries & certs", http.StatusInternalServerError)
+		log.Printf("sarif: %v, json: %v", err1, err2)
+		return
+	}
+
+	// Verify that the sarif results and json results are from the same repo and ref.
+	if sarifRepoPath != jsonRepoPath || sarifRepoRef != jsonRepoRef {
+		http.Error(w, "sarif and json results correspond to different repos/refs.", http.StatusInternalServerError)
+		return
+	}
+
+	repoPath := sarifRepoPath
+	repoRef := sarifRepoRef
+
+	// Split repo path into owner and repo name.
+	ownerName := repoPath[:strings.Index(repoPath, "/")]
+	repoName := repoPath[strings.Index(repoPath, "/")+1:]
+
+	// Verify that the repository and branch of the cert and request are equal.
+	reqRepo := r.Header["X-Repository"]
+	reqBranch := r.Header["X-Branch"]
+	if len(reqRepo) == 0 || len(reqBranch) == 0 || reqRepo[0] != repoPath || reqBranch[0] != repoRef {
+		http.Error(w, "repository and branch of cert doesn't match that of request", http.StatusInternalServerError)
+		return
+	}
+
+	// Make github client.
+	client := github.NewClient(nil)
+
+	// Get all workflows in the repository.
+	workflows, _, err := client.Actions.ListWorkflows(ctx, ownerName, repoName, nil)
+	if err != nil {
+		http.Error(w, "error listing workflows from repo", http.StatusInternalServerError)
+		log.Println(err)
+		return
+	}
+
+	wkflws := workflows.Workflows
+	wkflwPath := ""
+	for _, wkflw := range wkflws {
+		if *wkflw.Name == "Scorecards supply-chain security" {
```

Comment:
the `Name` may be `nil` if the users did not set the `name:` field in the workflow. Also, the name may change, we don't impose it. 

A better way: the certificate contains an `oid` with the `job_workflow_ref` and this will give you the path to the workflow associated with the certificate, see https://github.com/slsa-framework/slsa-verifier/blob/main/pkg/provenance.go#L368

Reply:
Not redundant: it has more information like filenames/line numbers. It's just a little hard to work with in general. But since we know which fields we populate, we should be able to parse it :-)

Reply:
Fixed. This is much better--thanks!